ci/bump_version.sh

```diff
@@ -15,50 +35,94 @@ echo "Searching for crates..."
 crates=($(find . -type d -exec bash -c '[ -f "$0"/Cargo.toml ]' '{}' \; -print))
 
 if [[ "${#crates[@]}" == "0" ]]; then
-  echo "No crates found."
+  echo -e "${RED}Error: No crates found."
   exit 1
 else
-  echo "Found ${#crates[@]} crate(s): ${crates[*]}"
+  echo -e "${CYAN}Found ${#crates[@]} crate(s): ${crates[*]}${NC}"
 fi
 
-file=Cargo.toml
 final_exit_code=0
 declare -a grep_failures
 declare -a lockfile_failures
 
-bump_version() {
-  # Catches every `version` that begins a line and doesn't end with a comma.
-  find_pattern='^version\s*=.*[^,]\s*$'
-  replace_pattern='s/'$find_pattern'/version = "'"$version"'"/'
+# Catches every `version` that begins a line and stores version inside double quotes
+find_pattern="^version = \".*\"$"
+replace_pattern="s/${find_pattern}/version = \"${version}\"/"
+
+get_current_version() {
+  local current_version=""
+  if [ "$(uname)" == "Darwin" ]; then
+    # macOS
+    current_version=$(grep -Eo "$find_pattern" "$file" | sed 's/^version = "//;s/"$//')
+  else
+    # Linux
+    current_version="$(grep -oP '(?<=^version = ")[^"]+' "$file" | head -n 1)"
+  fi
+
+  echo "$current_version"
+}
+
+replace_version() {
+  if [ "$(uname)" == "Darwin" ]; then
+    # macOS
+    sed -i '' "$replace_pattern" "$file"
+  else
+    # Linux
+    sed -i "$replace_pattern" "$file"
+  fi
+}
+
+find_and_replace() {
+  local crate=$1
+  local file="Cargo.toml"
 
-  grep -q "$find_pattern" "$file" && sed -i "$replace_pattern" "$file"
-  exit_code="$?"
-  if [[ "$exit_code" != "0" ]]; then
+  if grep -q "$find_pattern" "$file"; then
+    local prev_version=$(get_current_version)
+    replace_version
+    local new_version=$(get_current_version)
+    echo -e "${CYAN} Successfully changed the version inside $file for ${crate#./} (v$prev_version -> v$new_version)${NC}"
+  else
+    echo -e "${RED} Error: Failed to change the version inside $file for ${crate#./}${NC}"
     final_exit_code=1
-    grep_failures+=($1)
-    return
+    grep_failures+=("$crate")
+    return 1
   fi
+}
+
+update_lockfile() {
+  local crate=$1
+  local file="Cargo.lock"
 
-  cargo generate-lockfile
-  exit_code="$?"
-  if [[ "$exit_code" != "0" ]]; then
+  if cargo update --workspace; then
+    echo -e "${CYAN} Successfully updated $file for ${crate#./}${NC}"
+  else
+    echo -e "${RED} Error: Failed to update $file for ${crate#./}${NC}"
```

Comment:
I couldn't interpret this part ${crate#./}, particularly `#./`. Could you explain me what those three characters are supposed to do?

Reply:
It removes the prefix `./` from a string. For example, `./automap` will be printed as `automap` by using this formatting.